aten/src/ATen/TensorIterator.h

```diff
@@ -64,17 +73,40 @@ namespace internal {
 // no parallel algorithm (such as parallel_reduce) should split work into
 // smaller than GRAIN_SIZE chunks.
 constexpr int64_t GRAIN_SIZE = 32768;
+
+// Storage for a non-owning Tensor, without needing to include Tensor.h
+class TORCH_API OpaqueTensorRef {
```

Comment:
This looks a bit suspicious. If you're non-owning, why don't you just store TensorImpl*? You can reconstitute OptionalTensorRef from that without any trouble. The naming convention is also a bit suspect here.

Reply:
>If you're non-owning, why don't you just store TensorImpl*?

`TensorIterator::tensor()`, returns a `const Tensor&` so to maintain a compatible interface we need an honest to goodness `Tensor` object that outlives the member function call. So, it needs to be stored in the `TensorIterator`. It would be almost the same to return the tensor by value, but then we need to include `Tensor.h`, defeating the purpose of this PR.

> The naming convention is also a bit suspect here.

It is "Opaque" in the same sense that you can use `Tensor&` as an opaque type, i.e. without defining the `Tensor` class. It's a "Ref" in the sense of not incrementing the reference count.

Reply:
blegh ok

Reply:
Yeah, I agree it's ugly. I really wish I could just write `reinterpret_cast<const Tensor&>(op.tensor_base)` and it not be UB.

Reply:
Since you are getting an OptionalTensorRef out of this class, I'd probably prefer a wordier name like OpaqueOptionalTensorRef

Reply:
meh I guess I don't care too much about this